NDEV-3786 Use custom Solana Program (ALT-Updater) for strategy_stage_alt

artem-yazkov opened the pull request.

artem-yazkov committed:
NDEV-3786 ALT Updater program on strategy_stage_alt

artem-yazkov committed:
NDEV-3786 pass ALT System Program account to the ALT Updater program

NDEV-3786 use special compute units limit for the ALT Updater program

neonlabstech commented:
<details><summary>Dapps report</summary>

🔗[Cost report](https://github.com/neonlabsorg/neon-proxy.py/actions/runs/15752609096/artifacts/3361355225)

artem-yazkov committed:
NDEV-3786 some refactoring of class SolAltProg

artem-yazkov committed:
NDEV-3786 some refactoring of class SolAltProg

neonlabstech commented:
<details><summary>Dapps report</summary>

🔗[Cost report](https://github.com/neonlabsorg/neon-proxy.py/actions/runs/15756027134/artifacts/3362487372)

anton-lisanin commented on common/solana/alt_program.py:
```diff
@@ -51,14 +51,19 @@ def __eq__(self, other) -> bool:
+    # Program Pubkeys
     ID: Final[SolPubKey] = SolPubKey.from_raw(_alt.ID)
+    IDSystem: Final[SolPubKey] = SolPubKey.from_raw(_sys.ID)
+    IDALTUpdater: Final[SolPubKey] = SolPubKey.from_string('2opr1VoyXxpNePA4gcLBGPMPgrzgpyixuqDrE7EzKFWv')
```
Address should not be hardcoded.
Add it to the config.

artem-yazkov replied:
take it from `NEON_ALT_PROGRAM` environment variable

afalaleev commented on common/solana/alt_program.py:
```diff
@@ -51,14 +51,19 @@ def __eq__(self, other) -> bool:
 class SolAltProg:
+    # Program Pubkeys
     ID: Final[SolPubKey] = SolPubKey.from_raw(_alt.ID)
+    IDSystem: Final[SolPubKey] = SolPubKey.from_raw(_sys.ID)
```
Remove this line.

afalaleev commented on common/solana/alt_program.py:
```diff
@@ -51,14 +51,19 @@ def __eq__(self, other) -> bool:
     # CU limits for instructions
     CuLimitCreate: Final[int] = 25_000
     CuLimitExtend: Final[int] = 20_000
+    CuLimitCustom: Final[int] = 80_000
```
Why is it so big?
Did you try to run tests to get the real number?

afalaleev commented on common/solana/alt_program.py:
```diff
@@ -106,6 +111,23 @@ def make_extend_alt_ix(self, ident: SolAltID, account_key_list: Sequence[SolPubK
         )
 
+    def make_custom_alt_ix(self, ident: SolAltID, account_key_list: Sequence[SolPubKey]) -> SolTxIx:
+        accounts: List[SolAccountMeta] = [
```
Please keep the project naming:
```ix_acct_key_list: list[SolAccountMeta] .... ```

afalaleev commented on common/solana/alt_program.py:
```diff
@@ -106,6 +111,23 @@ def make_extend_alt_ix(self, ident: SolAltID, account_key_list: Sequence[SolPubK
+            SolAccountMeta(pubkey=ident.address, is_signer=False, is_writable=True),
+            SolAccountMeta(pubkey=self._payer, is_signer=True, is_writable=True),
+            SolAccountMeta(pubkey=self._payer, is_signer=True, is_writable=True),
+            SolAccountMeta(pubkey=self.IDSystem, is_signer=False, is_writable=False),
```
Replace with `SolSysProg.ID`

afalaleev commented on common/solana/alt_program.py:
```diff
@@ -106,6 +111,23 @@ def make_extend_alt_ix(self, ident: SolAltID, account_key_list: Sequence[SolPubK
+        accounts: List[SolAccountMeta] = [
+            SolAccountMeta(pubkey=ident.address, is_signer=False, is_writable=True),
+            SolAccountMeta(pubkey=self._payer, is_signer=True, is_writable=True),
+            SolAccountMeta(pubkey=self._payer, is_signer=True, is_writable=True),
```
We don't use a separate `payer` and `authority`.
Remove duplicate lines.

afalaleev commented on common/solana/alt_program.py:
```diff
@@ -106,6 +111,23 @@ def make_extend_alt_ix(self, ident: SolAltID, account_key_list: Sequence[SolPubK
             )
         )
 
+    def make_custom_alt_ix(self, ident: SolAltID, account_key_list: Sequence[SolPubKey]) -> SolTxIx:
```
Move to a separate Program -> SolExtAltProg

afalaleev replied:
Add ```assert len(account_key_list) > 0```

afalaleev commented on common/solana/alt_program.py:
```diff
@@ -51,14 +51,19 @@ def __eq__(self, other) -> bool:
+    # Program Pubkeys
     ID: Final[SolPubKey] = SolPubKey.from_raw(_alt.ID)
+    IDSystem: Final[SolPubKey] = SolPubKey.from_raw(_sys.ID)
+    IDALTUpdater: Final[SolPubKey] = SolPubKey.from_string('2opr1VoyXxpNePA4gcLBGPMPgrzgpyixuqDrE7EzKFWv')
```
Move to a separate program -> SolExtAltProg

artem-yazkov replied:
moved into separate class `SolExtAltProg`

afalaleev commented on common/solana/alt_program.py:
```diff
@@ -106,6 +111,23 @@ def make_extend_alt_ix(self, ident: SolAltID, account_key_list: Sequence[SolPubK
+            SolAccountMeta(pubkey=self.ID, is_signer=False, is_writable=False),
+        ]
+        for account_key in account_key_list:
+            accounts.append(SolAccountMeta(pubkey=account_key, is_signer=False, is_writable=False))
```
Remove `for` and replace with:
```
ix_acct_key_list.extend(map(lambda x: SolAccountMeta(pubkey=x, is_signer=False, is_writable=False), account_key_list))
```

afalaleev commented on common/solana_rpc/alt_builder.py:
```diff
@@ -150,6 +150,28 @@ def build_alt_tx_set(self, alt: SolAltInfo) -> SolAltTxSet:
 
         return SolAltTxSet(create_alt_tx_list=create_alt_tx_list, extend_alt_tx_list=extend_alt_tx_list)
 
+    def build_alt_custom_tx_set(self, alt: SolAltInfo) -> SolAltTxSet:
```
`def build_alt_tx_set()` isn't used anymore
You should comment or remove it, just because it will confuse the future developer - why does it exist but not get used?

afalaleev commented on common/solana/alt_program.py:
```diff
@@ -2,14 +2,14 @@
 import logging
 from enum import IntEnum
-from typing import Final, Sequence, Self
+from typing import Final, Sequence, List, Self
```
Remove `List`, it isn't required

artem-yazkov committed:
NDEV-3786 take ALT Updater Program ID from Environment variable

NDEV-3786 move ALT Updater wrapper logic into separate SolExtAltProg class

NDEV-3786 some improvements and cleanups

artem-yazkov committed:
NDEV-3786 remove old code based on SolAltProg creation and extend

artem-yazkov committed:
NDEV-3786 pass NEON_ALT_PROGRAM environment variable via compose files

afalaleev commented on common/solana/alt_program.py:
```diff
@@ -124,6 +128,32 @@ def make_close_alt_ix(self, ident: SolAltID) -> SolTxIx:
 
+class SolExtAltProg:
+    ID: Final[SolPubKey] = NEON_ALT_PROGRAM_ID
+    CuLimit: Final[int] = 65_000
```
Why is it so big?
Did you run tests to calculate the real numbers?

artem-yazkov replied:
@afalaleev I use test app to measure compute units for `ALT Updater`
https://github.com/neonlabsorg/alt-updater/blob/develop/client-rust/src/main.rs

In worst scenario, when new ALT created and 27 (`SolAltProg.MaxTxAccountCnt`) accounts inserted within, I got CUs value just under 52000

So, I use this value (52000) for now. Tests also show stable result at 52K Compute Units limit, and fails appears  with Limit < 50k

afalaleev replied:
I still don't understand why it's so big.

For example, ALT extending with 17-29 accounts < 12'000 CUs:
- https://explorer.solana.com/tx/5o7axNs2QMEgB1jVRVSNsLA3WXz1FkSPyJAKry1byRJHQQd7mfV2TUUdPR9pAnRbVp94c3UXUnnzTtdHaZURN9Hg
- https://explorer.solana.com/tx/5fZLtoatBiEd71thqPH2n85Uxxde2PiP2eggmgDTygMaFmNxSMfzP4TZnWpBZ5QpeV6vaVN1Gt2M3KsL64WhWpXt

ALT creating + ALT extending with 26 accounts < 25'000:
- https://explorer.solana.com/tx/5bxHWFp8AuWnMg2RVPUeLcAxZokF84NZw9mHQSBGbPVFzZ3Lg6DLv15tnddjMZZ39fBe376M3fF87dLwiBUpyxpJ

artem-yazkov replied:
I rewrote the ALT Updater Solana Program using [Pinocchio](https://github.com/anza-xyz/pinocchio), which greatly reduced CU overhead

I ran a series of tests, with the scenario: create ALT on an empty account and then extent it with 27 addresses (current limit from `SolAltProg.MaxTxAccountCnt`)

On most cases I got about 30K Compute Units for such tests.
CPI calls for `CreateLookupTable` and `ExtendLookupTable` takes about 24K CUs summary, so we got about 6K CUs overhead., spending mostly on `find_program_address` as well on instruction data preparation 

But some times I got some CUs consumption peaks to 34K for transaction
So, I set SolAltProg.CuLimit to 35000

@anton-lisanin @afalaleev FYI

afalaleev commented on common/solana_rpc/alt_builder.py:
```diff
@@ -109,46 +110,26 @@ async def rebuild_to_real_alt(self, alt: SolAltInfo) -> SolAltInfo:
-            create_alt_tx = SolLegacyTx(name=self._create_name, ix_list=ix_list)
-            create_alt_tx_list.append(create_alt_tx)
-
+    def build_ext_alt_tx_set(self, alt: SolAltInfo) -> SolAltTxSet:
```
What is the purpose to add `_ext_` in the name?
Should external modules know about the internal implementation?

artem-yazkov replied:
`ext` mean "based external ALT Solana program", as we use `SolExtAltProg` naming here.

can you suggest a better name?

afalaleev replied:
Just remove it -> `build_alt_tx_set`
There is no reason for external modules to be aware of the ALT module implementation.

afalaleev commented on common/solana_rpc/alt_builder.py:
```diff
@@ -109,46 +110,26 @@ async def rebuild_to_real_alt(self, alt: SolAltInfo) -> SolAltInfo:
 
         # List of txs to extend the Address Lookup Table
-        extend_alt_tx_list: list[SolLegacyTx] = list()
+        custom_alt_tx_list: list[SolLegacyTx] = list()
```
What is the purpose to add `custom_` prefix to the name?

artem-yazkov replied:
renamed to `alt_tx_list`

artem-yazkov committed:
NDEV-3786 minor refactoring

NDEV-3786 tune SolExtAltProg::CuLimit parameter

artem-yazkov committed:
NDEV-3786 tune SolAltProg::CuLimit to 52K

neonlabstech commented:
<details><summary>Dapps report</summary>

🔗[Cost report](https://github.com/neonlabsorg/neon-proxy.py/actions/runs/15896308436/artifacts/3408413576)

neonlabstech commented:
<details><summary><strong>⚠️ Cost Report Validation Failed ⚠️</strong></summary>

| dapp           | action                     | metric    |   DIFFERENCE |
|:---------------|:---------------------------|:----------|-------------:|
| aave           | borrow from lending pool   | acc_count |           28 |
| aave           | liquidation                | acc_count |           28 |
| aave           | repay                      | acc_count |           27 |
| aave           | repay                      | gas_used  |        12600 |
| aave           | withdraw                   | acc_count |           27 |
| saddle finance | swap usdc -> susd metapool | acc_count |           27 |

neonlabstech commented:
<details><summary>Dapps report</summary>

🔗[Cost report](https://github.com/neonlabsorg/neon-proxy.py/actions/runs/15899898671/artifacts/3409647670)

neonlabstech commented:
<details><summary><strong>⚠️ Cost Report Validation Failed ⚠️</strong></summary>

| dapp           | action                     | metric    |   DIFFERENCE |
|:---------------|:---------------------------|:----------|-------------:|
| aave           | borrow from lending pool   | acc_count |           27 |
| aave           | liquidation                | acc_count |           27 |
| aave           | repay                      | acc_count |           27 |
| aave           | withdraw                   | acc_count |           27 |
| saddle finance | swap usdc -> susd metapool | acc_count |           27 |

neonlabstech commented:
<details><summary>Solana Requests Statistics</summary>



| count   | min_time   | max_time   | average_time   | median_time   |
|---------|------------|------------|----------------|---------------|

</details>

neonlabstech commented:
<details><summary>Solana Requests Statistics</summary>



| count   | min_time   | max_time   | average_time   | median_time   |
|---------|------------|------------|----------------|---------------|

</details>

neonlabstech commented:
<details><summary><strong>⚠️ Cost Report Validation Failed ⚠️</strong></summary>

| dapp           | action                     | metric    |   DIFFERENCE |
|:---------------|:---------------------------|:----------|-------------:|
| aave           | borrow from lending pool   | acc_count |           27 |
| aave           | liquidation                | acc_count |           27 |
| aave           | repay                      | acc_count |           27 |
| aave           | withdraw                   | acc_count |           27 |
| saddle finance | swap usdc -> susd metapool | acc_count |           27 |

artem-yazkov committed:
NDEV-3786 address PR issues

neonlabstech commented:
<details><summary><strong>⚠️ Cost Report Validation Failed ⚠️</strong></summary>

| dapp           | action                     | metric    |   DIFFERENCE |
|:---------------|:---------------------------|:----------|-------------:|
| aave           | borrow from lending pool   | acc_count |           27 |
| aave           | liquidation                | acc_count |           26 |
| aave           | repay                      | acc_count |           27 |
| aave           | withdraw                   | acc_count |           27 |
| aave           | withdraw                   | trx_count |            2 |
| aave           | withdraw                   | gas_used  |        70400 |
| saddle finance | swap usdc -> susd metapool | acc_count |           27 |

neonlabstech commented:
<details><summary>Solana Requests Statistics</summary>



| count   | min_time   | max_time   | average_time   | median_time   |
|---------|------------|------------|----------------|---------------|

</details>

artem-yazkov committed:
NDEV-3786 use MaxTxAccountCnt constant from SolAltProg class

neonlabstech commented:
<details><summary>Dapps report</summary>

🔗[Cost report](https://github.com/neonlabsorg/neon-proxy.py/actions/runs/15921733665/artifacts/3417496813)

artem-yazkov committed:
NDEV-3786 reduce SolAltProg.CuLimit to 35k

neonlabstech commented:
<details><summary>Dapps report</summary>

🔗[Cost report](https://github.com/neonlabsorg/neon-proxy.py/actions/runs/16068198860/artifacts/3468320922)

neonlabstech commented:
<details><summary><strong>⚠️ Cost Report Validation Failed ⚠️</strong></summary>

| dapp           | action                     | metric    |   DIFFERENCE |
|:---------------|:---------------------------|:----------|-------------:|
| aave           | borrow from lending pool   | acc_count |           28 |
| aave           | liquidation                | acc_count |           28 |
| aave           | repay                      | acc_count |           28 |
| aave           | withdraw                   | acc_count |           28 |
| saddle finance | swap usdc -> susd metapool | acc_count |           28 |

neonlabstech commented:
<details><summary>Solana Requests Statistics</summary>



|                                   |   count |   min_time |   max_time |   average_time |   median_time |
|:----------------------------------|--------:|-----------:|-----------:|---------------:|--------------:|
| getAccountInfo                    |   45660 |      0     |      0.241 |    0.000491108 |        0      |
| getBlocks                         |   34006 |      0     |      0.169 |    0.000304917 |        0      |
| getBlock                          |   17156 |      0     |      0.241 |    0.000563826 |        0      |
| getTransaction                    |   12457 |      0     |      0.083 |    0.00251987  |        0.001  |
| getMultipleAccounts               |    8631 |      0     |      0.244 |    0.0013309   |        0.001  |
| sendTransaction                   |    5896 |      0     |      0.243 |    0.00541045  |        0.002  |
| getLatestBlockhash                |    5535 |      0     |      0.097 |    0.000442457 |        0      |
| getBalance                        |    1794 |      0     |      0.017 |    0.000255295 |        0      |
| getSignatureStatuses              |     441 |      0     |      0.005 |    0.000344671 |        0      |
| isBlockhashValid                  |     203 |      0     |      0.001 |    0.000187192 |        0      |
| getVersion                        |     180 |      0     |      0.011 |    0.000355556 |        0      |
| getSignaturesForAddress           |      90 |      0     |      0.002 |    0.000933333 |        0.001  |
| simulateTransaction               |      42 |      0     |      0.001 |    0.000333333 |        0      |
| requestAirdrop                    |      31 |      0     |      0.001 |    0.000322581 |        0      |
| getSlot                           |      25 |      0     |      0.002 |    0.00044     |        0      |
| getBlockHeight                    |      20 |      0     |      0.001 |    0.0002      |        0      |
| getTokenAccountBalance            |       3 |      0     |      0.001 |    0.000333333 |        0      |
| getFirstAvailableBlock            |       2 |      0     |      0     |    0           |        0      |
| getHealth                         |       2 |      0     |      0.001 |    0.0005      |        0.0005 |
| getMinimumBalanceForRentExemption |       1 |      0.001 |      0.001 |    0.001       |        0.001  |
| getProgramAccounts                |       1 |      0.053 |      0.053 |    0.053       |        0.053  |
| getClusterNodes                   |       1 |      0     |      0     |    0           |        0      |

</details>
